pkg/endpoint/bpf.go

```diff
@@ -655,87 +662,134 @@ func (e *Endpoint) regenerateBPF(owner Owner, currentDir, nextDir string, regenC
 	}
 
-	// Cache endpoint information
-	// TODO (ianvernon): why do we need to do this?
-	var epInfoCache *epInfoCache
-	if bpfHeaderfilesChanged {
-		epInfoCache = e.createEpInfoCache(nextDir)
+	// Cache endpoint information so that we can release the endpoint lock
+	// but still access information about the endpoint later if needed.
+	if datapathRegenCtx.bpfHeaderfilesChanged {
+		datapathRegenCtx.epInfoCache = e.createEpInfoCache(nextDir)
 	} else {
-		epInfoCache = e.createEpInfoCache(currentDir)
+		datapathRegenCtx.epInfoCache = e.createEpInfoCache(currentDir)
 	}
-	if epInfoCache == nil {
+	if datapathRegenCtx.epInfoCache == nil {
 		stats.prepareBuild.End(false)
-		e.Unlock()
-		err = fmt.Errorf("Unable to cache endpoint information")
-		return 0, compilationExecuted, err
+		return proxyCleanupFunc, fmt.Errorf("Unable to cache endpoint information")
 	}
-
 	// TODO: In Cilium v1.4 or later cycle, remove this.
 	os.RemoveAll(e.IPv6EgressMapPathLocked())
 	os.RemoveAll(e.IPv4EgressMapPathLocked())
 	os.RemoveAll(e.IPv6IngressMapPathLocked())
 	os.RemoveAll(e.IPv4IngressMapPathLocked())
+	return proxyCleanupFunc, nil
+}
+
+// regenerateBPF rewrites all headers and updates all BPF maps to reflect the
+// specified endpoint.
+// Must be called with endpoint.Mutex not held and endpoint.BuildMutex held.
+// Returns the policy revision number when the regeneration has called, a
+// boolean if the BPF compilation was executed and an error in case of an error.
+func (e *Endpoint) regenerateBPF(owner Owner, currentDir, nextDir string, regenContext *RegenerationContext) (revnum uint64, compiled bool, reterr error) {
+	stats := &regenContext.Stats
+	stats.waitingForLock.Start()
+
+	// Make sure that owner is not compiling base programs while we are
+	// regenerating an endpoint.
+	owner.GetCompilationLock().RLock()
+	defer owner.GetCompilationLock().RUnlock()
+
+	// Cache endpoint ID for use later.
+	e.UnconditionalRLock()
+	epID := e.StringID()
+	e.RUnlock()
+
+	bpfRegenContext := newDatapathRegenerationContext()
+	defer bpfRegenContext.completionCancel()
 
-	e.Unlock()
+	// Wrap logic which requires the endpoint to be locked within a function
+	// & call within regenerateBPF so that endpoint lock release can be deferred
+	// within the function. If this isn't done, and something else panics
+	// while the endpoint regeneration lock is held, and a deferred function
+	// within regenerateBPF tries to grab the endpoint lock, it will deadlock
+	// because the lock is already held, which obfuscates what happened in the
+	// initial panic.
+	// Returns an error, as well as a function which may need deferring
+	// based once this function returns. This function is intentionally not
+	// deferred in this function itself because it may acquire the endpoint
+	// lock.
+	revFunc, err := e.runPreCompilationSteps(owner, bpfRegenContext, currentDir, nextDir, regenContext)
+
+	// Note that revFunc may hold the endpoint lock.
+	if revFunc != nil {
```

Comment:
I wonder if it would be safer to make the contract of `runPreCompilationSteps` always return a non-nil function, but return a dummy on error.

Reply:
Only thing I can argue against that is that we have to perform an unnecessary function call, but I do like protecting against nil-dereferencing (which would cause a panic ;) ).

Reply:
I've changed this to be a function which is performed upon endpoint with two parameters, a `RegenerationContext` and an `error`. Let me know if you prefer how I've done it.